packages/tracing/src/browser/browsertracing.ts

```diff
@@ -0,0 +1,241 @@
+    };
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public setupOnce(_: (callback: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    BrowserTracing._getCurrentHub = getCurrentHub;
+
+    if (!global || !global.location) {
+      return;
+    }
+
+    // TODO: is it fine that this is mutable operation? Could also do = [...routingInstr, setHeaderContext]?
+    BrowserTracing.options.routingInstrumentationProcessors.push(setHeaderContext);
+    BrowserTracing._initRoutingInstrumentation();
+  }
+
+  /** Start routing instrumentation */
+  private static _initRoutingInstrumentation(): void {
+    const { startTransactionOnPageLoad, startTransactionOnLocationChange } = BrowserTracing.options;
+
+    if (startTransactionOnPageLoad) {
+      BrowserTracing._activeTransaction = BrowserTracing._createRouteTransaction('pageload');
+    }
+
+    let startingUrl: string | undefined = global.location.href;
+
+    addInstrumentationHandler({
+      callback: ({ to, from }: { to: string; from?: string }) => {
+        /**
+         * This early return is there to account for some cases where navigation transaction
+         * starts right after long running pageload. We make sure that if from is undefined
+         * and that a valid startingURL exists, we don't uncessarily create a navigation transaction.
+         *
+         * This was hard to duplicate, but this behaviour stopped as soon as this fix
+         * was applied. This issue might also only be caused in certain development environments
+         * where the usage of a hot module reloader is causing errors.
+         */
+        if (from === undefined && startingUrl && startingUrl.indexOf(to) !== -1) {
+          startingUrl = undefined;
+          return;
+        }
+        if (startTransactionOnLocationChange && from !== to) {
+          startingUrl = undefined;
+          if (BrowserTracing._activeTransaction) {
+            // We want to finish all current ongoing idle transactions as we
+            // are navigating to a new page.
+            BrowserTracing._activeTransaction.finishIdleTransaction();
+          }
+          BrowserTracing._activeTransaction = BrowserTracing._createRouteTransaction('navigation');
+        }
+      },
+      type: 'history',
+    });
+  }
+
+  /** Create pageload/navigation idle transaction. */
+  private static _createRouteTransaction(op: 'pageload' | 'navigation'): IdleTransaction | undefined {
+    if (!BrowserTracing._getCurrentHub) {
+      return undefined;
+    }
+
+    const { beforeNavigate, idleTimeout, routingInstrumentationProcessors } = BrowserTracing.options;
+
+    // if beforeNavigate returns undefined, we should not start a transaction.
+    const name = beforeNavigate(global.location);
+    if (name === undefined) {
+      return undefined;
+    }
+
+    let context: TransactionContext = { name, op };
+    if (routingInstrumentationProcessors) {
+      for (const processor of routingInstrumentationProcessors) {
```

Comment:
My only concern here is how expensive this operation could be.

Reply:
It should be fine, we already do it for event processors and didn't have any perf issues.
I'd think about handling broken processors though.